finetune/association.py

```diff
@@ -0,0 +1,353 @@
+        :param **kwargs: key-value pairs of config items to override.
+        """
+        self.multi_label = False
+        super().__init__(config=config, **kwargs)
+
+    def _get_input_pipeline(self):
+        return AssociationPipeline(config=self.config, multi_label=False)
+
+    def _initialize(self):
+        if self.config.multi_label_sequences:
+            raise FinetuneError("Multi label association not supported")
+        return super()._initialize()
+
+    def finetune(self, Xs, Y=None, batch_size=None):
+        """
+        :param Xs: A list of strings.
+        :param Y: A list of labels of the same format as sequence labeling but with an option al additional field
+        of the form:
+        ```
+            {
+                ...
+                "association":{
+                        "index": a,
+                        "relationship": relationship_name
+                }
+                ...
+        ```
+        where index is the index of the relationship target into the label list and relationship_name is the type of
+        the relationship.
+        """
+        if self.config.possible_associations is None:
+            raise FinetuneError("Please set config.possible_associations before calling finetune.")
+        Xs, Y_new, association_type, association_idx, idxs = indico_to_finetune_sequence(
+            Xs, labels=Y, multi_label=False, none_value="<PAD>"
+        )
+
+        Y = list(zip(Y_new, association_type, association_idx, idxs)) if Y is not None else None
+        return super().finetune(Xs, Y=Y, batch_size=batch_size)
+
+    def prune_probs(self, prob_matrix, labels):
+        viable_edges = self.config.viable_edges
+        possible_associations = list(self.input_pipeline.association_encoder.classes_)
+        if viable_edges is None:
+            return prob_matrix
+        for i, l1 in enumerate(labels):
+            if l1 not in viable_edges:
+                prob_matrix[i, :, :] = 0.0
+
+            elif None not in viable_edges[l1]:
+                prob_matrix[i, :, self.input_pipeline.association_pad_idx] = 0.0
+            for cls in possible_associations:
+                for j, l2 in enumerate(labels):
+                    if l1 not in viable_edges or l2 not in [c_t[0] for c_t in viable_edges[l1] if
+                                                            c_t and c_t[1] == cls]:
+                        prob_matrix[i, j, possible_associations.index(cls)] = 0.0  # this edge doesnt fit the schema
+        return prob_matrix
+
+    def predict(self, X):
+        """
+        Produces a list of most likely class labels as determined by the fine-tuned model.
+
+        :param X: A list / array of text, shape [batch]
+        :returns: list of class labels.
+        """
+        LOGGER.warning("config.viable_edges is not set, this is probably incorrect.")
+
+        #TODO(Ben) combine this into the sequence labeling model??
+
+        chunk_size = self.config.max_length - 2
```

Comment:
Even if we don't factor everything out we probably should look into factoring out the chunking logic.

Reply:
I think this is something I wanted to discuss, I don't think chunking makes any sense in the case of the association model. So I wanted to remove it entirely, but i'm not sure whether having the restricted maximum separation of associations is something that is okay? So I left it in.